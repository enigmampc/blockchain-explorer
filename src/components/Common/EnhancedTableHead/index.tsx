import React from 'react'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import TableSortLabel from '@material-ui/core/TableSortLabel'
import styled from 'styled-components'

export enum FlexAlign {
  'flexStart' = 'flex-start',
  'flexEnd' = 'flex-end',
  'center' = 'center',
}

export type HeaderCell = {
  id: string
  useClassShowOnDesktop: boolean
  align: FlexAlign
  label: string
}

export interface EnhancedTableHeadProps extends React.HTMLAttributes<HTMLDivElement> {
  headerCells: HeaderCell[]
  order: 'asc' | 'desc'
  orderBy: string
  onRequestSort: CallableFunction
}

const TableHeadText = styled.span`
  color: #999;
  font-size: 12px;
  white-space: nowrap;

  & + svg {
    opacity: 1;
  }
`

const EnhancedTableHead = ({ headerCells, order, orderBy, onRequestSort }: EnhancedTableHeadProps) => {
  const createSortHandler = (property: string) => (event: React.SyntheticEvent) => {
    onRequestSort(event, property)
  }

  return (
    <TableHead>
      <TableRow>
        {headerCells.map(headerCell => {
          const active = orderBy === headerCell.id

          return (
            <TableCell
              key={headerCell.id}
              sortDirection={active ? order : false}
              className={headerCell.useClassShowOnDesktop ? 'showOnDesktop' : ''}
              style={{
                paddingRight: '5px',
                display: headerCell.useClassShowOnDesktop ? 'none' : 'table-cell',
              }}
            >
              {active ? (
                <TableSortLabel
                  active={active}
                  direction={order}
                  hideSortIcon={false}
                  onClick={createSortHandler(headerCell.id)}
                  style={{
                    justifyContent: headerCell.align,
                    display: 'flex',
                  }}
                >
                  <TableHeadText>{headerCell.label}</TableHeadText>
                </TableSortLabel>
              ) : (
                <TableHeadText
                  onClick={createSortHandler(headerCell.id)}
                  style={{
                    justifyContent: headerCell.align,
                    display: 'flex',
                    cursor: 'pointer',
                  }}
                >
                  {headerCell.label}
                </TableHeadText>
              )}
            </TableCell>
          )
        })}
      </TableRow>
    </TableHead>
  )
}

export default EnhancedTableHead
