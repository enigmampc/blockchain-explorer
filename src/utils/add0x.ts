const add0x = (hex: string): string => (hex ? (hex.indexOf('0x') === 0 ? hex : `0x${hex}`) : '')

export default add0x
