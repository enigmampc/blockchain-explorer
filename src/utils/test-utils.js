import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import theme from '../theme'

const getStyledComponents = StyledComponent => {
  const componentClass = StyledComponent().type.styledComponentId
  return document.getElementsByClassName(componentClass)
}

const history = createBrowserHistory()

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Router history={history}>{children}</Router>
  </ThemeProvider>
)

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'

export { customRender as render, getStyledComponents, history }
