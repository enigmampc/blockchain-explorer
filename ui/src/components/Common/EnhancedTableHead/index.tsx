import React, { createRef } from 'react'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import ArrowDownwardIcon from '@material-ui/icons/ArrowDownward'
import { FilterSVG } from './img/FilterSVG'
import FilterItem from '../FilterItem'
import { Button } from '../Button'
import styled from 'styled-components'

export enum FlexAlign {
  'center' = 'center',
  'end' = 'flex-end',
  'start' = 'flex-start',
}

export type FilterOption = {
  title: string
  value: string
}

export type HeaderCell = {
  align: FlexAlign
  id: string
  label: string
  filter?: boolean
  filterOptions?: FilterOption[]
  sortable: boolean
  useClassShowOnDesktop: boolean
}

export interface EnhancedTableHeadProps extends React.HTMLAttributes<HTMLDivElement> {
  headerCells: HeaderCell[]
  order: 'asc' | 'desc'
  orderBy: string
  onRequestSort: CallableFunction
  onRequestFilter?: CallableFunction
  filteredDataSet?: boolean
}

const COLOR_DEFAULT = '#999'
const COLOR_ACTIVE = '#333'

const TableHeadWrapper = styled.span<{ alignElements?: string }>`
  align-items: center;
  display: flex;
  justify-content: ${props => (props.alignElements ? props.alignElements : 'flex-start')};
`

const TableHeadText = styled.span<{ active?: boolean }>`
  color: ${props => (props.active ? COLOR_ACTIVE : COLOR_DEFAULT)};
  font-size: 12px;
  white-space: nowrap;
`

const FilterWrapper = styled.div<{ active?: boolean }>`
  background-color: transparent;
  border: none;
  cursor: pointer;
  margin: 0 0 0 5px;
  outline: none;
  padding: 0;
  position: relative;

  > svg {
    fill: ${COLOR_DEFAULT};
  }

  &:focus-within {
    > div {
      display: block;
    }

    > svg {
      fill: ${COLOR_ACTIVE};
    }
  }
`

const FilterDropdown = styled.div`
  background-color: #fff;
  border-radius: 3px;
  border: solid 1px ${props => props.theme.borders.borderColor};
  box-shadow: 0 0 15px 0 rgba(0, 0, 0, 0.02);
  display: none;
  left: 50%;
  position: absolute;
  top: calc(100% + 10px);
  transform: translateX(-50%);
  width: 125px;

  &::before {
    border-bottom: 5px solid #fff;
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    content: '';
    height: 5px;
    left: 50%;
    position: absolute;
    top: -5px;
    transform: translateX(-50%);
    width: 5px;
    z-index: 5;
  }

  &::after {
    border-bottom: 7px solid ${props => props.theme.borders.borderColor};
    border-left: 7px solid transparent;
    border-right: 7px solid transparent;
    content: '';
    height: 7px;
    left: 50%;
    position: absolute;
    top: -7px;
    transform: translateX(-50%);
    width: 7px;
    z-index: 1;
  }
`

const FilterButtonContainer = styled.span`
  display: flex;
  padding: 6px 8px;
`

const HelperFocusItem = styled.span`
  height: 0;
  line-height: 0;
  width: 0;
  position: absolute;
  z-index: -123456;
`

const ButtonStyled = styled(Button)`
  border-radius: 2px;
  flex-grow: 1;
  font-size: 12px;
  font-weight: 600;
  height: 22px;
  line-height: 22px;
  text-transform: uppercase;
`

export type TableFilterProps = {
  options?: FilterOption[]
  myRef: any
  applyFilter: Function
  item: string
}

const TableFilter = ({ options, myRef, applyFilter, item }: TableFilterProps) => {
  options = options || []
  const [optionsSelection, setOptionsSelection] = React.useState(
    options.reduce((acc: { [index: string]: boolean }, { value }) => {
      acc[value] = false
      return acc
    }, {}),
  )

  const onChangeFilterItem = (item: string, isChecked: boolean) => {
    setOptionsSelection(prevState => ({ ...prevState, [item]: isChecked }))
  }

  return (
    <>
      <HelperFocusItem tabIndex={-1} ref={myRef} />
      <FilterWrapper tabIndex={-1}>
        <FilterSVG />
        <FilterDropdown>
          {options &&
            options.map(({ title, value }) => (
              <FilterItem onChange={onChangeFilterItem} text={title} value={value} key={value} />
            ))}
          <FilterButtonContainer>
            <ButtonStyled onClick={() => applyFilter(optionsSelection, item)} backgroundColor="#e72e9d">
              Apply
            </ButtonStyled>
          </FilterButtonContainer>
        </FilterDropdown>
      </FilterWrapper>
    </>
  )
}

const EnhancedTableHead = ({
  headerCells,
  order,
  orderBy,
  onRequestSort,
  onRequestFilter,
  filteredDataSet,
}: EnhancedTableHeadProps) => {
  const createSortHandler = (property: string) => (event: React.SyntheticEvent) => {
    onRequestSort(event, property)
  }

  const myRef = createRef<HTMLDivElement>()

  const applyFilter = (optionsSelection: { [index: string]: boolean }, item: string) => {
    // apply filter
    const selectedOptions = Object.keys(optionsSelection).filter(k => optionsSelection[k])
    onRequestFilter && onRequestFilter(item, selectedOptions)

    // close dropdown
    if (myRef.current) {
      myRef.current.focus()
    }
  }

  return (
    <TableHead>
      <TableRow>
        {headerCells.map(headerCell => {
          const activeSorting = orderBy === headerCell.id
          const sortProps: any = {
            direction: activeSorting ? 'asc' : 'desc',
            onClick: createSortHandler(headerCell.id),
            style: {
              cursor: 'pointer',
              fill: activeSorting ? COLOR_ACTIVE : COLOR_DEFAULT,
              height: '16px',
              marginLeft: '8px',
              transform: order === 'desc' ? 'rotate(180deg)' : '',
              width: '16px',
            },
          }

          return (
            <TableCell
              className={headerCell.useClassShowOnDesktop ? 'showOnDesktop' : ''}
              key={headerCell.id}
              style={{
                display: headerCell.useClassShowOnDesktop ? 'none' : 'table-cell',
                paddingRight: '5px',
              }}
            >
              <TableHeadWrapper alignElements={headerCell.align}>
                <TableHeadText active={activeSorting}>{headerCell.label}</TableHeadText>
                {headerCell.sortable ? <ArrowDownwardIcon {...sortProps} /> : null}
                {headerCell.filter ? (
                  <TableFilter
                    myRef={myRef}
                    options={headerCell.filterOptions}
                    item={headerCell.id}
                    applyFilter={applyFilter}
                  />
                ) : null}
              </TableHeadWrapper>
            </TableCell>
          )
        })}
      </TableRow>
    </TableHead>
  )
}

export default EnhancedTableHead
