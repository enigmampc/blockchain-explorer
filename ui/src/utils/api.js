import { formatNumber } from './numeralFormats'

export const cryptoCompare = async () => {
  const url = new URL('https://min-api.cryptocompare.com/data/pricemultifull')
  const params = {
    fsyms: 'ENG',
    tsyms: 'USD',
  }
  url.search = new URLSearchParams(params).toString()
  const {
    RAW: {
      ENG: {
        USD: { PRICE, MKTCAP },
      },
    },
  } = await (await fetch(url)).json().catch(console.error)

  return {
    price: formatNumber(PRICE, 4),
    marketCap: formatNumber(MKTCAP, 0),
  }
}

export const coinMarketCap = async () => {
  const url = new URL('https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest')
  const params = {
    id: 2044,
  }
  url.search = new URLSearchParams(params).toString()
  const headers = new Headers({
    'X-CMC_PRO_API_KEY': process.env.REACT_APP_COINMARKET_CAP_API_KEY,
  })
  const init = {
    method: 'GET',
    headers,
    cache: 'default',
  }
  const request = new Request(url, init)
  const response = await (await fetch(request)).json()

  if (response.error_code) throw new Error(response.error_message)

  const {
    data: {
      [params.id]: {
        quote: {
          USD: { market_cap, price },
        },
      },
    },
  } = response

  return {
    price: formatNumber(price, 4),
    marketCap: formatNumber(market_cap, 0),
  }
}

export const binance = async () => {
  const binanceURL = 'https://api.binance.com/api/v1/ticker/price?symbol='
  const { price } = await (await fetch(`${binanceURL}ENGBTC`)).json().catch()

  return {
    price: formatNumber(price, 4),
    marketCap: '',
  }
}

export const enigmaTicker = async () => {
  const enigmaAPIURL = `${process.env.REACT_APP_ENIGMA_API}/ticker`
  const response = await fetch(enigmaAPIURL, { mode: 'no-cors' })
  const ticker = { priceUsd: '0', marketCapUsd: '0' }

  if (response.ok) {
    const { price_usd, market_cap_usd } = await response.json()
    ticker.priceUsd = price_usd
    ticker.marketCapUsd = market_cap_usd
  }

  return {
    price: formatNumber(+ticker.price_usd, 4),
    marketCap: formatNumber(+ticker.market_cap_usd, 0),
  }
}
