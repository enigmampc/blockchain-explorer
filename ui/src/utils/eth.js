import Web3 from 'web3'
import getNetworkDetailsBy from './networks'

export const ENG_DECIMALS = 8

class EthAPI {
  constructor() {
    // web3
    if (process.env.REACT_APP_ETH_NETWORK_ID) {
      const { url } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)
      this.web3 = new Web3(new Web3.providers.HttpProvider(url))
    } else {
      this.web3 = new Web3(new Web3.providers.HttpProvider(process.env.REACT_APP_ETH_URL))
    }
  }

  getBlockNumber() {
    return this.web3.eth.getBlockNumber()
  }

  async getBlockTimestamp(blockNumber) {
    const block = await this.web3.eth.getBlock(blockNumber)
    return (block && block.timestamp) || '0'
  }

  /**
   * Request timestamp for all the blocks in a list of block numbers
   * @param { number[] } range
   * @returns {Promise<number[]>}
   */
  async getBatchBlocksTimestamps(range) {
    const batch = new this.web3.BatchRequest()

    const blocks = range.map(blockNumber => {
      return new Promise((resolve, reject) => {
        const request = this.web3.eth.getBlock.request(blockNumber, (error, data) => {
          if (error) {
            reject(error)
          } else {
            resolve(data)
          }
        })
        batch.add(request)
      })
    })

    batch.execute()

    return (await Promise.all(blocks)).map(({ timestamp }) => timestamp)
  }
}

export default new EthAPI()
