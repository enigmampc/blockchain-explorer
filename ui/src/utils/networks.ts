// list of ethNetworks provided on eth-cli
import ethNetworks from './networks.json'

const UNKNOWN_NAME = 'unknown'
const UNKNOWN_ID = -1

const byId = (Object.keys(ethNetworks) as Array<keyof typeof ethNetworks>).reduce(
  (acc: { [index: string]: any }, key: keyof typeof ethNetworks) => {
    acc[ethNetworks[key].id] = {
      name: key,
      url: ethNetworks[key].url,
      label: ethNetworks[key].label,
    }
    return acc
  },
  {},
)

const getNetworkDetailsById = (id: number) => byId[id] || byId[UNKNOWN_ID]

const getNetworkDetailsByName = (name: keyof typeof ethNetworks) => ethNetworks[name] || ethNetworks[UNKNOWN_NAME]

const getNetworkDetailsBy = (by: string): Function => {
  if (by === 'id') {
    return getNetworkDetailsById
  }
  if (by === 'name') {
    return getNetworkDetailsByName
  }

  throw new Error('invalid selection type')
}

export default getNetworkDetailsBy
