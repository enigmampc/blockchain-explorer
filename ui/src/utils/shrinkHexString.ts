const shrinkHexString = (value = '', start = 1, end = 1) => {
  const matcher = new RegExp(`(?:(0?x?[a-f0-9]{${start}})).*([a-f0-9]{${end}})`, 'i')
  const [, ...matches] = value.match(matcher) || []
  return matches.join('...')
}

export default shrinkHexString
