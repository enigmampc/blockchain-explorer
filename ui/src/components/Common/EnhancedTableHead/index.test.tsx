import React from 'react'
import { fireEvent, render } from '../../../utils/test-utils'
import EnhancedTableHead, { HeaderCell, FlexAlign } from './'
import Table from '@material-ui/core/Table'

const HEADER_CELLS: HeaderCell[] = [
  { id: '_id', useClassShowOnDesktop: false, align: FlexAlign.start, label: 'Task ID' },
  { id: 'secretContract', useClassShowOnDesktop: true, align: FlexAlign.start, label: 'Secret Contact Address' },
  {
    id: 'workerAddress',
    useClassShowOnDesktop: true,
    align: FlexAlign.start,
    label: 'Worker Address (Ethereum)',
  },
  { id: 'status', useClassShowOnDesktop: false, align: FlexAlign.center, label: 'Status' },
  { id: 'sentOn', useClassShowOnDesktop: true, align: FlexAlign.center, label: 'Sent On' },
  { id: 'completedOn', useClassShowOnDesktop: true, align: FlexAlign.center, label: 'Completed On' },
]

it('renders EnhancedTableHead', () => {
  // Given
  const headerCells = HEADER_CELLS
  const order = 'asc'
  const orderBy = 'status'
  const onRequestSort = jest.fn()

  // When
  const { container } = render(
    <Table>
      <EnhancedTableHead headerCells={headerCells} order={order} orderBy={orderBy} onRequestSort={onRequestSort} />
    </Table>,
  )

  // Then
  expect(container).toMatchSnapshot()
})

it('renders EnhancedTableHead without title', () => {
  // Given
  const headerCells = HEADER_CELLS
  const order = 'asc'
  const orderBy = 'status'
  const onRequestSort = jest.fn()

  // When
  const { getByText } = render(
    <Table>
      <EnhancedTableHead headerCells={headerCells} order={order} orderBy={orderBy} onRequestSort={onRequestSort} />
    </Table>,
  )
  const byText = getByText(/completed on/i)
  fireEvent.click(byText)

  // Then
  expect(onRequestSort).toHaveBeenCalled()
})
