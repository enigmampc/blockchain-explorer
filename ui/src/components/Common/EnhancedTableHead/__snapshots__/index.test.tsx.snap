// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EnhancedTableHead 1`] = `
.c0 {
  color: #999;
  font-size: 12px;
  white-space: nowrap;
}

.c0 + svg {
  opacity: 1;
}

<div>
  <table
    class="MuiTable-root"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: table-cell;"
        >
          <span
            class="c0"
            style="justify-content: flex-start; display: flex; cursor: pointer;"
          >
            Task ID
          </span>
        </th>
        <th
          class="MuiTableCell-root showOnDesktop MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: none;"
        >
          <span
            class="c0"
            style="justify-content: flex-start; display: flex; cursor: pointer;"
          >
            Secret Contact Address
          </span>
        </th>
        <th
          class="MuiTableCell-root showOnDesktop MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: none;"
        >
          <span
            class="c0"
            style="justify-content: flex-start; display: flex; cursor: pointer;"
          >
            Worker Address (Ethereum)
          </span>
        </th>
        <th
          aria-sort="ascending"
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: table-cell;"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
            role="button"
            style="justify-content: center; display: flex;"
            tabindex="0"
          >
            <span
              class="c0"
            >
              Status
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          class="MuiTableCell-root showOnDesktop MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: none;"
        >
          <span
            class="c0"
            style="justify-content: center; display: flex; cursor: pointer;"
          >
            Sent On
          </span>
        </th>
        <th
          class="MuiTableCell-root showOnDesktop MuiTableCell-head"
          scope="col"
          style="padding-right: 5px; display: none;"
        >
          <span
            class="c0"
            style="justify-content: center; display: flex; cursor: pointer;"
          >
            Completed On
          </span>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
